crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -51,15 +51,30 @@ fn all_narrowing_constraints_for_pattern<'db>(
     db: &'db dyn Db,
     pattern: PatternConstraint<'db>,
 ) -> NarrowingConstraints<'db> {
-    NarrowingConstraintsBuilder::new(db, Constraint::Pattern(pattern)).finish()
+    NarrowingConstraintsBuilder::new(
+        db,
+        Constraint {
+            node: ConstraintNode::Pattern(pattern),
+            negative: false,
+        },
+    )
+    .finish()
 }
 
 #[salsa::tracked(return_ref)]
 fn all_narrowing_constraints_for_expression<'db>(
     db: &'db dyn Db,
     expression: Expression<'db>,
+    negative: bool,
```

Comment:
Hmm, I think Salsa does something kind of inefficient when a tracked function has multiple inputs; it internally interns all the arguments. Although it's slightly weird-looking, I think we will get more efficient results here if we break into two separate Salsa tracked functions, `all_narrowing_constraints_for_expression` and `all_negative_narrowing_constraints_for_expression`.